### ISSUE TITLE
(VI_1) entering FAULT state (no IPv4 address for interface) - even though Interface has IP

### ISSUE THREAD

**Issue by lorenzPowedale** [acassen/keepalived#2091]
**Describe the bug**
When Keepalived starts I get an error message "(VI_1) entering FAULT state (no IPv4 address for interface)". The problem is that I have configured that Interface. There is probably something wrong with my conf file. Please help me, thanks in advance.

**To Reproduce**
Im using Ubuntu 20.04.

Interface Config:
```
# This is the network config written by 'subiquity'
network:
  ethernets:
    eno1:
      dhcp4: true
    eno2:
      dhcp4: true
    eno3:
      dhcp4: no
      dhcp6: no
      addresses: [10.20.20.204/24]
      nameservers:
        addresses: [8.8.8.8,8.8.8.4]
    eno4:
      dhcp4: no
      dhcp6: no
      addresses: [10.20.20.205/24]
      nameservers:
        addresses: [8.8.8.8,8.8.8.4]
  version: 2
  renderer: networkd
```
```
3: eno4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 24:6e:96:22:66:0d brd ff:ff:ff:ff:ff:ff
    inet 10.20.20.205/24 brd 10.20.20.255 scope global eno4
       valid_lft forever preferred_lft forever
    inet6 fe80::266e:96ff:fe22:660d/64 scope link 
       valid_lft forever preferred_lft forever
```

Keepalived Config:
```
vrrp_instance VI_1 {
    state MASTER
    interface eno4
    virtual_router_id 101
    priority 101
    advert_int 1
    authentication {
        auth_type PASS
        auth_pass 1111
    }
    virtual_ipaddress {
        10.20.20.205/24
    }
}

```
Slave:
```
vrrp_instance VI_1 {
    state MASTER
    interface eno4
    virtual_router_id 101
    priority 100
    advert_int 1
    authentication {
        auth_type PASS
        auth_pass 1111
    }
    virtual_ipaddress {
        10.20.20.205/24
    }
}
```

Keepalived status:
```
● keepalived.service - Keepalive Daemon (LVS and VRRP)
     Loaded: loaded (/lib/systemd/system/keepalived.service; enabled; vendor preset: enabled)
     Active: active (running) since Tue 2022-02-01 18:57:02 UTC; 6s ago
   Main PID: 3350 (keepalived)
      Tasks: 2 (limit: 154493)
     Memory: 1.9M
     CGroup: /system.slice/keepalived.service
             ├─3350 /usr/sbin/keepalived --dont-fork
             └─3354 /usr/sbin/keepalived --dont-fork

Feb 01 18:57:02 inginiousserver1 Keepalived[3350]: WARNING - keepalived was build for newer Linux 5.4.151, running on Linux 5.4.0-96-generic #109-Ubuntu SMP Wed Jan 12 16:49:16 UTC 2022
Feb 01 18:57:02 inginiousserver1 Keepalived[3350]: Command line: '/usr/sbin/keepalived' '--dont-fork'
Feb 01 18:57:02 inginiousserver1 Keepalived[3350]: Opening file '/etc/keepalived/keepalived.conf'.
Feb 01 18:57:02 inginiousserver1 Keepalived[3350]: Starting VRRP child process, pid=3354
Feb 01 18:57:02 inginiousserver1 Keepalived_vrrp[3354]: Registering Kernel netlink reflector
Feb 01 18:57:02 inginiousserver1 Keepalived_vrrp[3354]: Registering Kernel netlink command channel
Feb 01 18:57:02 inginiousserver1 Keepalived_vrrp[3354]: Opening file '/etc/keepalived/keepalived.conf'.
Feb 01 18:57:02 inginiousserver1 Keepalived_vrrp[3354]: (VI_1) entering FAULT state (no IPv4 address for interface)
Feb 01 18:57:02 inginiousserver1 Keepalived_vrrp[3354]: (VI_1) entering FAULT state
Feb 01 18:57:02 inginiousserver1 Keepalived_vrrp[3354]: Registering gratuitous ARP shared channel
```

**Expected behavior**
Keepalived should start correctly.

The Server with the "Master"-Config should get the IP assigned
The Server with the "Slave"-Config should not get the IP assigned, until the Server with the "Master"-Config fails.

**Keepalived version**
```
Keepalived v2.0.19 (10/19,2019)
```

**Distro:**
 - Ubuntu
 - 20.04.3
 - x86_64

**Details of any containerisation or hosted service (e.g. AWS)**
None

**Configuration file:**
Master:
```
vrrp_instance VI_1 {
    state MASTER
    interface eno4
    virtual_router_id 101
    priority 101
    advert_int 1
    authentication {
        auth_type PASS
        auth_pass 1111
    }
    virtual_ipaddress {
        10.20.20.205/24
    }
}

```
Slave:
```
vrrp_instance VI_1 {
    state MASTER
    interface eno4
    virtual_router_id 101
    priority 100
    advert_int 1
    authentication {
        auth_type PASS
        auth_pass 1111
    }
    virtual_ipaddress {
        10.20.20.205/24
    }
}
```

**Notify and track scripts**
```
None
```

**System Log entries**
```
● keepalived.service - Keepalive Daemon (LVS and VRRP)
     Loaded: loaded (/lib/systemd/system/keepalived.service; enabled; vendor preset: enabled)
     Active: active (running) since Tue 2022-02-01 18:57:02 UTC; 6s ago
   Main PID: 3350 (keepalived)
      Tasks: 2 (limit: 154493)
     Memory: 1.9M
     CGroup: /system.slice/keepalived.service
             ├─3350 /usr/sbin/keepalived --dont-fork
             └─3354 /usr/sbin/keepalived --dont-fork

Feb 01 18:57:02 inginiousserver1 Keepalived[3350]: WARNING - keepalived was build for newer Linux 5.4.151, running on Linux 5.4.0-96-generic #109-Ubuntu SMP Wed Jan 12 16:49:16 UTC 2022
Feb 01 18:57:02 inginiousserver1 Keepalived[3350]: Command line: '/usr/sbin/keepalived' '--dont-fork'
Feb 01 18:57:02 inginiousserver1 Keepalived[3350]: Opening file '/etc/keepalived/keepalived.conf'.
Feb 01 18:57:02 inginiousserver1 Keepalived[3350]: Starting VRRP child process, pid=3354
Feb 01 18:57:02 inginiousserver1 Keepalived_vrrp[3354]: Registering Kernel netlink reflector
Feb 01 18:57:02 inginiousserver1 Keepalived_vrrp[3354]: Registering Kernel netlink command channel
Feb 01 18:57:02 inginiousserver1 Keepalived_vrrp[3354]: Opening file '/etc/keepalived/keepalived.conf'.
Feb 01 18:57:02 inginiousserver1 Keepalived_vrrp[3354]: (VI_1) entering FAULT state (no IPv4 address for interface)
Feb 01 18:57:02 inginiousserver1 Keepalived_vrrp[3354]: (VI_1) entering FAULT state
Feb 01 18:57:02 inginiousserver1 Keepalived_vrrp[3354]: Registering gratuitous ARP shared channel
```

**Comment by pqarmitage** [comment]
@lorenzPowedale There are a number of issues here.

1. You have configured the 'Slave' with `state MASTER`. Since you want this instance to start in backup state, it would be more sensible to configure `state BACKUP`. However, my recommendation is to delete the `state ...` statements on both 'Master' and 'Slave' - they serve no useful purpose.
2. The virtual_ip_address you have configured is the address that is configured on eno4. Keepalived cannot send adverts using a virtual_ip_address, and so when it is considering the addresses on the interface to use, it has to ignore any addresses that are also VIPs (or eVIPs). As a consequence eno4 has no addresses that keepalived can use for sending adverts, and it therefore goes to fault state. If you manually execute `ip addr add 10.20.20.200 brd + dev eno4` for example (make sure the address you use is not already used on the network), then keepalived will then see that there is an address that it can use for adverts and will transition to backup, and then master state.
3. When keepalived starts, it removes any addresses specified in `virtual_ip_address` or `virtual_ip_address_excluded` blocks, since it assumes that they must have been added by a previously running instance of keepalived which failed to shutdown cleanly. This means that when keepalived starts, it deletes the 10.20.20.205/24 IP address from eno4 (if you run keepalived with the `-D` option, you should see a log message `(VI_1) removing VIPs.` indicating this.
4. The system is configuring an all-1s broadcast address for eno4 (i.e. 10.20.20.255, most easily specified with `brd +`), whereas your keepalived configuration specifies an all-0s broadcast address (i.e. 10.20.20.0, which can be specified by `brd -`), because it is the default. To rectify this you should change the VIP specification to `10.20.20.205/24 brd +`.
5. I am not clear why you are configuring eno3 and eno4 on the same subnet, but using different network interfaces. You might get better performance if you added eno3 and eno4 to a bond interface, and configured both 10.20.20.204 and 10.20.20.205 on the bond interface, providing of course that your network infrastructure supports bonding.

In summary:
1. you need to have an IP address configured on eno4 that is different from any keepalived virtual_ip_addresses, and make sure that your system doesn't configure any addresses on eno4 that are also VIPs.
2. you need to correct the broadcast address (i.e. add `brd +`) for the VIP.

**Comment by lorenzPowedale** [comment]
Thank you very much!!!

I was under the assumption that I needed a separate Interface just for the Virtual-IP. That's the reason why I configured 2 IPs from the same subnet on seperate Interfaces.

The only thing I needed to change was:
```
vrrp_instance VI_1 {
    state MASTER
    interface eno4 -> eno3
    virtual_router_id 101
    priority 101
    advert_int 1
    authentication {
        auth_type PASS
        auth_pass 1111
    }
    virtual_ipaddress {
        10.20.20.205/24
    }
}
```

**Comment by liudonghua123** [comment]
I have encountered the same problems, I fixed it via setting a ip for the interface or use an interface which has an an ip already. In the latter, the vip will assigned as a secondary ip.

Tips: `ifconfig` will now show the secondary ip, you need to use `ip addr show`.
